obj: add pmemobj_mem[cpy|set|move] with flags

marcinslusarz opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmdk/2769)
<!-- Reviewable:end -->

pbalcer commented on src/libpmem/aarch64/init.c:
```diff
@@ -51,7 +51,8 @@ memmove_nodrain_libc(void *pmemdest, const void *src, size_t len,
 	(void) flags;
```
not needed anymore.

marcinslusarz replied:
Fixed.

pbalcer commented on src/libpmem/aarch64/init.c:
```diff
@@ -51,7 +51,8 @@ memmove_nodrain_libc(void *pmemdest, const void *src, size_t len,
 
 	memmove(pmemdest, src, len);
-	pmem_flush(pmemdest, len);
+	if (!(flags & PMEM_MEM_NOFLUSH))
```
maybe we could have a global macro:
`pmem_flush_flags(pmemdest, len, flags)` ?
Would allow you to move the flags checking to a single function.

marcinslusarz replied:
Done.

krzycz commented:
Reviewed 33 of 35 files at r1, 4 of 4 files at r2, 2 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[doc/libpmem/pmem_memmove_persist.3.md, line 122 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IY3e6_sK4HA3PoQ1O:-L8IY3e6_sK4HA3PoQ1P:b2ngrbt) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/doc/libpmem/pmem_memmove_persist.3.md#L122)):*
> ```Markdown
> 
> + **PMEM_MEM_NOFLUSH** - Don't flush anything. This implies **PMEM_MEM_NODRAIN**.
>   Using this flag only makes sense when it's followed by **pmem_flush** or
> ```

(3)

---

*[doc/libpmem/pmem_memmove_persist.3.md, line 123 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IY4sBBKFO9qcptDvZ:-L8IY4sBBKFO9qcptDv_:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/doc/libpmem/pmem_memmove_persist.3.md#L123)):*
> ```Markdown
> + **PMEM_MEM_NOFLUSH** - Don't flush anything. This implies **PMEM_MEM_NODRAIN**.
>   Using this flag only makes sense when it's followed by **pmem_flush** or
>   **pmem_persist**.
> ```

''

---

*[doc/libpmemobj/pmemobj_memcpy_persist.3.md, line 115 at r4](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8ZIycu67kup-Z7X_oz:-L8ZIycu67kup-Z7X_p-:b94zpdd) ([raw file](https://github.com/pmem/pmdk/blob/ea1e786e2c98d7025a05919685ca38c829f4219f/doc/libpmemobj/pmemobj_memcpy_persist.3.md#L115)):*
> ```Markdown
> The **pmemobj_memmove**(), **pmemobj_memcpy**() and **pmemobj_memset**() functions
> provide the same memory copying as their namesakes **memmove**(3), **memcpy**(3),
> and **memset**(3), and ensure that the result has been flushed to persistence
> ```

Hmmm, is it always true? I mean, if there is PMEM_MEM_NOFLUSH in the flags, then nothing is flushed, right? Correct me if I'm wrong.

---

*[doc/libpmemobj/pmemobj_memcpy_persist.3.md, line 116 at r4](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8ZJASnSUSxzfed8b7C:-L8ZJASnSUSxzfed8b7D:bfcurnx) ([raw file](https://github.com/pmem/pmdk/blob/ea1e786e2c98d7025a05919685ca38c829f4219f/doc/libpmemobj/pmemobj_memcpy_persist.3.md#L116)):*
> ```Markdown
> provide the same memory copying as their namesakes **memmove**(3), **memcpy**(3),
> and **memset**(3), and ensure that the result has been flushed to persistence
> before returning. Flags have the same meaning as in **pmem_memmove**(),
> ```

pmem_memmove(3), ...

---

*[src/include/libpmemobj/base.h, line 189 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IVCszsXRYtEzy8NVH:-L8IVCt-KYdnQXjlFCMw:b-dbehv8) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/include/libpmemobj/base.h#L189)):*
> ```objc
> 
> /*
>  * Pmemobj version of memcpy. Data copied is made persistent.
> ```

Is made or is not - it depends on the flags, I suppose.

---

*[src/libpmemobj/obj.c, line 360 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IWD3h95x1AWRxyAPA:-L8IWD3h95x1AWRxyAPB:b-e2y8q0) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L360)):*
> ```C
> 	LOG(15, "dest %p src %p len %zu flags 0x%x", dest, src, len, flags);
> 
> 	pmem_memcpy(dest, src, len, PMEM_MEM_NOFLUSH);
> ```

Is it better than regular memcpy in case of non-PMEM?

---

*[src/libpmemobj/obj.c, line 373 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IWHRQ6fSVa3SjtqRA:-L8IWHRQ6fSVa3SjtqRB:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L373)):*
> ```C
> 	LOG(15, "dest %p src %p len %zu flags 0x%x", dest, src, len, flags);
> 
> 	pmem_memmove(dest, src, len, PMEM_MEM_NOFLUSH);
> ```

''

---

*[src/libpmemobj/obj.c, line 386 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IWSuKL6ThxaN4wKCl:-L8IWSuKL6ThxaN4wKCm:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L386)):*
> ```C
> 	LOG(15, "dest %p c 0x%02x len %zu flags 0x%x", dest, c, len, flags);
> 
> 	pmem_memset(dest, c, len, PMEM_MEM_NOFLUSH);
> ```

''

---

*[src/test/obj_mem/TEST0, line 38 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IX4o_KYV36gRe1jPd:-L8IX4oa9rGxVNKbtFEe:blzug90) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0#L38)):*
> ```
> # and pmemobj_memset
> #
> export UNITTEST_NAME=obj_mem/TEST0
> ```

I thought we don't need this anymore...

---

*[src/test/obj_mem/TEST0, line 45 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IXWJWVCznhA6u8Sbo:-L8IXWJWVCznhA6u8Sbp:b3v5ojz) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0#L45)):*
> ```
> 
> require_test_type medium
> ```

If TEST1 requires valgrind (memcheck), we can explicitly disable it here, so if somebody would run tests with `-m force-enable` the same test won't be run twice.

---

*[src/test/obj_mem/TEST0.PS1, line 42 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IXOKEuPbxEAiiQCVA:-L8IXOKEuPbxEAiiQCVB:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0.PS1#L42)):*
> ```
>     $DIR = ""
>     )
> $Env:UNITTEST_NAME = "obj_mem/TEST0"
> ```

''

---

*[src/test/obj_mem/TEST1, line 38 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IXQwfMA-K-nCHZ_dc:-L8IXQwfMA-K-nCHZ_dd:b32v1h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST1#L38)):*
> ```
> # and pmemobj_memset
> #
> export UNITTEST_NAME=obj_mem/TEST1
> ```

''

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2769)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 32 of 39 files reviewed at latest revision, 14 unresolved discussions.

---

*[doc/libpmem/pmem_memmove_persist.3.md, line 122 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IY3e6_sK4HA3PoQ1O:-L8ao5OU32QPo1kVHAFt:b-scc20h) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/doc/libpmem/pmem_memmove_persist.3.md#L122)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

(3)
</blockquote></details>

I realized that this list is too short, so I replaced with generic description.

---

*[doc/libpmem/pmem_memmove_persist.3.md, line 123 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IY4sBBKFO9qcptDvZ:-L8aoOIuEDflFUFc34Nu:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/doc/libpmem/pmem_memmove_persist.3.md#L123)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---

*[doc/libpmemobj/pmemobj_memcpy_persist.3.md, line 115 at r4](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8ZIycu67kup-Z7X_oz:-L8arN5v5YBdK9dC58u4:bt2vr1b) ([raw file](https://github.com/pmem/pmdk/blob/ea1e786e2c98d7025a05919685ca38c829f4219f/doc/libpmemobj/pmemobj_memcpy_persist.3.md#L115)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Hmmm, is it always true? I mean, if there is PMEM_MEM_NOFLUSH in the flags, then nothing is flushed, right? Correct me if I'm wrong.
</blockquote></details>

Fixed.

---

*[doc/libpmemobj/pmemobj_memcpy_persist.3.md, line 116 at r4](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8ZJASnSUSxzfed8b7C:-L8arOGtAsPOFV-FRznJ:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/ea1e786e2c98d7025a05919685ca38c829f4219f/doc/libpmemobj/pmemobj_memcpy_persist.3.md#L116)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

pmem_memmove(3), ...
</blockquote></details>

Done.

---

*[src/include/libpmemobj/base.h, line 189 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IVCszsXRYtEzy8NVH:-L8b4PeO34wm8PNRswYi:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/include/libpmemobj/base.h#L189)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is made or is not - it depends on the flags, I suppose.
</blockquote></details>

Done.

---

*[src/libpmemobj/obj.c, line 360 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IWD3h95x1AWRxyAPA:-L8ap5PxClxgKI6hiuNA:b-xcyjzp) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L360)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Is it better than regular memcpy in case of non-PMEM?
</blockquote></details>

It's about atomicity. If application crashes in the middle of libc memcpy we can't guarantee correctness.

---

*[src/libpmemobj/obj.c, line 373 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IWHRQ6fSVa3SjtqRA:-L8apNgsBbBjHQsVs4ZZ:b-89a4l7) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L373)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Ditto

---

*[src/libpmemobj/obj.c, line 386 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IWSuKL6ThxaN4wKCl:-L8apRMg1UZFceHqVWvR:b-89a4l7) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L386)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Ditto

---

*[src/test/obj_mem/TEST0, line 38 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IX4o_KYV36gRe1jPd:-L8apwfpAuiKkr6NW5Re:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0#L38)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

I thought we don't need this anymore...
</blockquote></details>

Done.

---

*[src/test/obj_mem/TEST0, line 45 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IXWJWVCznhA6u8Sbo:-L8aq8zA26bcXJfg1_32:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0#L45)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

If TEST1 requires valgrind (memcheck), we can explicitly disable it here, so if somebody would run tests with `-m force-enable` the same test won't be run twice.
</blockquote></details>

Done.

---

*[src/test/obj_mem/TEST0.PS1, line 42 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IXOKEuPbxEAiiQCVA:-L8aqAJY95VlplfjLY2_:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST0.PS1#L42)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---

*[src/test/obj_mem/TEST1, line 38 at r2](https://beta.reviewable.io/reviews/pmem/pmdk/2769#-L8IXQwfMA-K-nCHZ_dc:-L8aqC9v2ujzI0560dza:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/test/obj_mem/TEST1#L38)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

''
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/pmem/pmdk/2769)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemobj/obj.c, line 360 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IWD3h95x1AWRxyAPA:-L8bFEx-S8mUXJqfKzw1:bymhdc7) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L360)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

It's about atomicity. If application crashes in the middle of libc memcpy we can't guarantee correctness.
</blockquote></details>

Perhaps it would be good to add such comment in the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2769#-:-L8bFL6MHMpHdmDijAZU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 38 of 39 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemobj/obj.c, line 360 at r2](https://reviewable.io:443/reviews/pmem/pmdk/2769#-L8IWD3h95x1AWRxyAPA:-L8bJvz8_Eoo365zihQI:b-896fix) ([raw file](https://github.com/pmem/pmdk/blob/da1756b8c8a27679c2107d5142af7372ca1664e4/src/libpmemobj/obj.c#L360)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

Perhaps it would be good to add such comment in the code.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmdk/2769)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/pmem/pmdk/2769)*
<!-- Sent from Reviewable.io -->
